jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java

```diff
@@ -78,7 +80,16 @@ public boolean checkOrigin(String originHeaderValue)
     @Override
     public List<Extension> getNegotiatedExtensions(List<Extension> installed, List<Extension> requested)
     {
-        return requested;
+        List<Extension> negotiatedExtensions = new ArrayList<>();
+        for (Extension ext : requested)
+        {
+            // Only choose the first extension if multiple with the same name.
+            long matches = negotiatedExtensions.stream().filter(e -> e.getName().equals(ext.getName())).count();
+            if (matches == 0)
+                negotiatedExtensions.add(ext);
+        }
+
+        return negotiatedExtensions;
```

Comment:
This isn't right.

A few points ...

* Multiple copies of the same extension are allowed. Eg: `identity; step=1, permessage-deflate, identity; step=2`
* This API should validate each `requested` name against the list in `installed` and not return `requested` that does not show up in `installed`.

Reply:
The `permessage-deflate` extension has special language that it reserves the RSV1 bit.
So if you have a offered extensions of `identity; step=1, permessage-deflate, identity; step=2, permessage-deflate; client_no_context_takeover`, then what happens is ...

1. `identity; step=1` is initialized successfully.
2. `permessage-deflate` is initialized, and reserves the RSV1 bit. successfully.
3. `identity; step=2` is initialized successfully.
4. `permessage-deflate; client_no_context_takeover` is initialized, and _fails_ to reserve the RSV1 bit, throwing an exception, and erroring out the handshake. (the language used in RFC6455 is `MUST immediately _Fail the WebSocket Connection_`)

Reply:
I don't think this is how it works, I'm not so sure with the `identity` extension case, but I know that with `permessage-deflate` the client can offer the same extension multiple times with different parameters and the server can choose the one to use. It would not result in a failed handshake.

[RFC6455 Section 9.1](https://tools.ietf.org/html/rfc6455#section-9.1)
```
Any interactions
between multiple extensions MAY be defined in the documents defining
the extensions.  In the absence of such definitions, the
interpretation is that the header fields listed by the client in its
request represent a preference of the header fields it wishes to use,
with the first options listed being most preferable.
```

[RFC7692 'Compression Extensions' Section 5](https://tools.ietf.org/html/rfc7692#section-5)
```
A client may also offer multiple
PMCE choices to the server by including multiple elements in the
"Sec-WebSocket-Extensions" header, one for each PMCE offered.  This
set of elements MAY include multiple PMCEs with the same extension
name to offer the possibility to use the same algorithm with
different configuration parameters.
```